{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9389c3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bf50101a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"pipeline2/downloads/homology_databases/homo_sapiens.homologies.tsv.gz\", compression=\"gzip\", delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2dc98c7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>protein_stable_id</th>\n",
       "      <th>species</th>\n",
       "      <th>identity</th>\n",
       "      <th>homology_type</th>\n",
       "      <th>homology_gene_stable_id</th>\n",
       "      <th>homology_protein_stable_id</th>\n",
       "      <th>homology_species</th>\n",
       "      <th>homology_identity</th>\n",
       "      <th>dn</th>\n",
       "      <th>ds</th>\n",
       "      <th>goc_score</th>\n",
       "      <th>wga_coverage</th>\n",
       "      <th>is_high_confidence</th>\n",
       "      <th>homology_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSP00000480818</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>27.4336</td>\n",
       "      <td>ortholog_many2many</td>\n",
       "      <td>ENSNLEG00000033214</td>\n",
       "      <td>ENSNLEP00000009514</td>\n",
       "      <td>nomascus_leucogenys</td>\n",
       "      <td>72.5752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54878896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSP00000480818</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>20.9861</td>\n",
       "      <td>ortholog_many2many</td>\n",
       "      <td>ENSNLEG00000028011</td>\n",
       "      <td>ENSNLEP00000042954</td>\n",
       "      <td>nomascus_leucogenys</td>\n",
       "      <td>66.6667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>75.36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54878908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSP00000480818</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>24.6523</td>\n",
       "      <td>ortholog_one2many</td>\n",
       "      <td>ENSPTRG00000000232</td>\n",
       "      <td>ENSPTRP00000058531</td>\n",
       "      <td>pan_troglodytes</td>\n",
       "      <td>80.5785</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.69</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55560202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSP00000480818</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>14.5386</td>\n",
       "      <td>ortholog_many2many</td>\n",
       "      <td>ENSCHOG00000007713</td>\n",
       "      <td>ENSCHOP00000006838</td>\n",
       "      <td>choloepus_hoffmanni</td>\n",
       "      <td>18.6992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.74</td>\n",
       "      <td>0.0</td>\n",
       "      <td>61939974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSP00000480818</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>12.1365</td>\n",
       "      <td>ortholog_many2many</td>\n",
       "      <td>ENSCHOG00000007714</td>\n",
       "      <td>ENSCHOP00000006839</td>\n",
       "      <td>choloepus_hoffmanni</td>\n",
       "      <td>16.9312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>61939989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_stable_id protein_stable_id       species  identity  \\\n",
       "0  ENSG00000271254   ENSP00000480818  homo_sapiens   27.4336   \n",
       "1  ENSG00000271254   ENSP00000480818  homo_sapiens   20.9861   \n",
       "2  ENSG00000271254   ENSP00000480818  homo_sapiens   24.6523   \n",
       "3  ENSG00000271254   ENSP00000480818  homo_sapiens   14.5386   \n",
       "4  ENSG00000271254   ENSP00000480818  homo_sapiens   12.1365   \n",
       "\n",
       "        homology_type homology_gene_stable_id homology_protein_stable_id  \\\n",
       "0  ortholog_many2many      ENSNLEG00000033214         ENSNLEP00000009514   \n",
       "1  ortholog_many2many      ENSNLEG00000028011         ENSNLEP00000042954   \n",
       "2   ortholog_one2many      ENSPTRG00000000232         ENSPTRP00000058531   \n",
       "3  ortholog_many2many      ENSCHOG00000007713         ENSCHOP00000006838   \n",
       "4  ortholog_many2many      ENSCHOG00000007714         ENSCHOP00000006839   \n",
       "\n",
       "      homology_species  homology_identity  dn  ds  goc_score  wga_coverage  \\\n",
       "0  nomascus_leucogenys            72.5752 NaN NaN        0.0          4.03   \n",
       "1  nomascus_leucogenys            66.6667 NaN NaN        0.0         75.36   \n",
       "2      pan_troglodytes            80.5785 NaN NaN        0.0          6.69   \n",
       "3  choloepus_hoffmanni            18.6992 NaN NaN        0.0         38.74   \n",
       "4  choloepus_hoffmanni            16.9312 NaN NaN        0.0         12.31   \n",
       "\n",
       "   is_high_confidence  homology_id  \n",
       "0                 0.0     54878896  \n",
       "1                 0.0     54878908  \n",
       "2                 0.0     55560202  \n",
       "3                 0.0     61939974  \n",
       "4                 0.0     61939989  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e314922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hmm = pd.read_csv(\"pipeline3/config/hmm_table.tsv.gz\", compression=\"gzip\", delimiter=\"\\t\",names = [\"family\",\"gene_id\",\"sequence_id\", \"species\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e06cee7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>sequence_id</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTHR11432_SF3</td>\n",
       "      <td>ENSTNIG00000004376</td>\n",
       "      <td>ENSTNIP00000007010</td>\n",
       "      <td>tetraodon_nigroviridis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PTHR22773_SF41</td>\n",
       "      <td>ENSTNIG00000004377</td>\n",
       "      <td>ENSTNIP00000007011</td>\n",
       "      <td>tetraodon_nigroviridis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTHR10422</td>\n",
       "      <td>ENSTNIG00000000533</td>\n",
       "      <td>ENSTNIP00000003835</td>\n",
       "      <td>tetraodon_nigroviridis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PTHR22888</td>\n",
       "      <td>ENSTNIG00000000112</td>\n",
       "      <td>ENSTNIP00000002153</td>\n",
       "      <td>tetraodon_nigroviridis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TF656662</td>\n",
       "      <td>ENSTNIG00000001181</td>\n",
       "      <td>ENSTNIP00000003540</td>\n",
       "      <td>tetraodon_nigroviridis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           family             gene_id         sequence_id  \\\n",
       "0   PTHR11432_SF3  ENSTNIG00000004376  ENSTNIP00000007010   \n",
       "1  PTHR22773_SF41  ENSTNIG00000004377  ENSTNIP00000007011   \n",
       "2       PTHR10422  ENSTNIG00000000533  ENSTNIP00000003835   \n",
       "3       PTHR22888  ENSTNIG00000000112  ENSTNIP00000002153   \n",
       "4        TF656662  ENSTNIG00000001181  ENSTNIP00000003540   \n",
       "\n",
       "                  species  \n",
       "0  tetraodon_nigroviridis  \n",
       "1  tetraodon_nigroviridis  \n",
       "2  tetraodon_nigroviridis  \n",
       "3  tetraodon_nigroviridis  \n",
       "4  tetraodon_nigroviridis  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "333d3aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = pd.read_csv(\"pipeline3/config/Valid_species_new_way.txt\", delimiter=\" \", names = [\"species\",\"species_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6de57c2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>species_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "      <td>ENSAMEG0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>amphiprion_percula</td>\n",
       "      <td>ENSAPEG0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>astatotilapia_calliptera</td>\n",
       "      <td>ENSACLG0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>astyanax_mexicanus</td>\n",
       "      <td>ENSAMXG0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bos_taurus</td>\n",
       "      <td>ENSBTAG0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species species_id\n",
       "0    ailuropoda_melanoleuca   ENSAMEG0\n",
       "1        amphiprion_percula   ENSAPEG0\n",
       "2  astatotilapia_calliptera   ENSACLG0\n",
       "3        astyanax_mexicanus   ENSAMXG0\n",
       "4                bos_taurus   ENSBTAG0"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "79fc929c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# consider only those homologous species in our species list\n",
    "df = df[df.homology_species.isin(species.species)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fbe0684c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample 100 orthologs from each species, with replacement in case less than 100 orthologous pairs present\n",
    "ortholog_samples = df[df.homology_type.str.contains(\"ortholog\")].groupby(\"homology_species\").sample(100,replace=True)\n",
    "ortholog_samples = ortholog_samples.drop_duplicates() # drop the duplicated samples from replacement"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeadafbb",
   "metadata": {},
   "source": [
    "## Generating gene pairs in the same protein family between two species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d4b505b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison_species = \"ailuropoda_melanoleuca\"\n",
    "temp_hmm  = hmm[hmm.species.isin([\"homo_sapiens\",comparison_species])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3803aef5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>sequence_id</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3544665</th>\n",
       "      <td>PTHR26451_SF247</td>\n",
       "      <td>ENSAMEG00000005298</td>\n",
       "      <td>ENSAMEP00000041297</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3544669</th>\n",
       "      <td>PTHR26451_SF319</td>\n",
       "      <td>ENSAMEG00000027833</td>\n",
       "      <td>ENSAMEP00000019703</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3544671</th>\n",
       "      <td>PTHR22754</td>\n",
       "      <td>ENSAMEG00000025036</td>\n",
       "      <td>ENSAMEP00000025763</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3544673</th>\n",
       "      <td>PTHR11006_SF52</td>\n",
       "      <td>ENSAMEG00000031399</td>\n",
       "      <td>ENSAMEP00000043819</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3544683</th>\n",
       "      <td>PTHR22754</td>\n",
       "      <td>ENSAMEG00000001136</td>\n",
       "      <td>ENSAMEP00000001303</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3786593</th>\n",
       "      <td>PTHR19271</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>ENSP00000354554</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3786594</th>\n",
       "      <td>PTHR12231_SF9</td>\n",
       "      <td>ENSG00000277666</td>\n",
       "      <td>ENSP00000482263</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3786595</th>\n",
       "      <td>PTHR24118_SF46</td>\n",
       "      <td>ENSG00000276760</td>\n",
       "      <td>ENSP00000478796</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3786596</th>\n",
       "      <td>PTHR26451_SF53</td>\n",
       "      <td>ENSG00000275249</td>\n",
       "      <td>ENSP00000484075</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3786597</th>\n",
       "      <td>PTHR26451_SF97</td>\n",
       "      <td>ENSG00000274792</td>\n",
       "      <td>ENSP00000480318</td>\n",
       "      <td>homo_sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38268 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  family             gene_id         sequence_id  \\\n",
       "3544665  PTHR26451_SF247  ENSAMEG00000005298  ENSAMEP00000041297   \n",
       "3544669  PTHR26451_SF319  ENSAMEG00000027833  ENSAMEP00000019703   \n",
       "3544671        PTHR22754  ENSAMEG00000025036  ENSAMEP00000025763   \n",
       "3544673   PTHR11006_SF52  ENSAMEG00000031399  ENSAMEP00000043819   \n",
       "3544683        PTHR22754  ENSAMEG00000001136  ENSAMEP00000001303   \n",
       "...                  ...                 ...                 ...   \n",
       "3786593        PTHR19271     ENSG00000198727     ENSP00000354554   \n",
       "3786594    PTHR12231_SF9     ENSG00000277666     ENSP00000482263   \n",
       "3786595   PTHR24118_SF46     ENSG00000276760     ENSP00000478796   \n",
       "3786596   PTHR26451_SF53     ENSG00000275249     ENSP00000484075   \n",
       "3786597   PTHR26451_SF97     ENSG00000274792     ENSP00000480318   \n",
       "\n",
       "                        species  \n",
       "3544665  ailuropoda_melanoleuca  \n",
       "3544669  ailuropoda_melanoleuca  \n",
       "3544671  ailuropoda_melanoleuca  \n",
       "3544673  ailuropoda_melanoleuca  \n",
       "3544683  ailuropoda_melanoleuca  \n",
       "...                         ...  \n",
       "3786593            homo_sapiens  \n",
       "3786594            homo_sapiens  \n",
       "3786595            homo_sapiens  \n",
       "3786596            homo_sapiens  \n",
       "3786597            homo_sapiens  \n",
       "\n",
       "[38268 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_hmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0bd89988",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the list of genes that\n",
    "family_pairs = []\n",
    "for family in temp_hmm.family.drop_duplicates():\n",
    "    # get all genes sharing the family\n",
    "    temp = temp_hmm[temp_hmm.family == family]\n",
    "    # split this into two databases for the human and the comparison\n",
    "    human = temp[temp.species == \"homo_sapiens\"]\n",
    "    other= temp[temp.species == comparison_species]\n",
    "    # skip if only present in one species\n",
    "    if (human.shape[0] == 0) | (other.shape[0] == 0):\n",
    "        continue\n",
    "    # enumerate all pairs of gene between the species, in the same family\n",
    "    pairs = pd.Series(itertools.product(list(human.gene_id),list(other.gene_id)))\n",
    "    # add it to the list\n",
    "    family_pairs.append(pairs)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "343aebe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate this into a series of all gene pairs sharing a family\n",
    "family_pairs = pd.concat(family_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0fcfa4b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    (ENSG00000183024, ENSAMEG00000005298)\n",
       "0    (ENSG00000172150, ENSAMEG00000027833)\n",
       "1    (ENSG00000172150, ENSAMEG00000028317)\n",
       "2    (ENSG00000172146, ENSAMEG00000027833)\n",
       "3    (ENSG00000172146, ENSAMEG00000028317)\n",
       "                     ...                  \n",
       "1    (ENSG00000198908, ENSAMEG00000019140)\n",
       "0    (ENSG00000198088, ENSAMEG00000010461)\n",
       "0    (ENSG00000080572, ENSAMEG00000010454)\n",
       "0    (ENSG00000101901, ENSAMEG00000011941)\n",
       "0    (ENSG00000169684, ENSAMEG00000012716)\n",
       "Length: 65289, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "family_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ac811942",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all orthologous pairs\n",
    "ortholog_pairs = df[(df.homology_species == comparison_species) & (df.homology_type.str.contains(\"ortholog\"))][[\"gene_stable_id\",\"homology_gene_stable_id\"]].apply(tuple, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1d4f8e6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16710"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "family_pairs.isin(ortholog_pairs).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a6eb5aa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "149        (ENSG00000276017, ENSAMEG00000027505)\n",
       "378        (ENSG00000277196, ENSAMEG00000011835)\n",
       "654        (ENSG00000278633, ENSAMEG00000027505)\n",
       "707        (ENSG00000274847, ENSAMEG00000007524)\n",
       "1087       (ENSG00000276345, ENSAMEG00000002153)\n",
       "                           ...                  \n",
       "3982737    (ENSG00000182899, ENSAMEG00000019159)\n",
       "3982933    (ENSG00000185621, ENSAMEG00000004187)\n",
       "3983644    (ENSG00000198899, ENSAMEG00000025297)\n",
       "3984843    (ENSG00000277666, ENSAMEG00000014514)\n",
       "3985208    (ENSG00000275249, ENSAMEG00000026002)\n",
       "Length: 18702, dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ortholog_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c215c6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,j in ortholog_pairs[~ortholog_pairs.isin(family_pairs)]:\n",
    "    print(df[(df.gene_stable_id == i) & (df.homology_gene_stable_id == j)].homology_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0ab5a47f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>protein_stable_id</th>\n",
       "      <th>species</th>\n",
       "      <th>identity</th>\n",
       "      <th>homology_type</th>\n",
       "      <th>homology_gene_stable_id</th>\n",
       "      <th>homology_protein_stable_id</th>\n",
       "      <th>homology_species</th>\n",
       "      <th>homology_identity</th>\n",
       "      <th>dn</th>\n",
       "      <th>ds</th>\n",
       "      <th>goc_score</th>\n",
       "      <th>wga_coverage</th>\n",
       "      <th>is_high_confidence</th>\n",
       "      <th>homology_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9044</th>\n",
       "      <td>ENSG00000088782</td>\n",
       "      <td>ENSP00000371825</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>65.6566</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>ENSAMEG00000011292</td>\n",
       "      <td>ENSAMEP00000011868</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "      <td>65.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83864229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_stable_id protein_stable_id       species  identity  \\\n",
       "9044  ENSG00000088782   ENSP00000371825  homo_sapiens   65.6566   \n",
       "\n",
       "         homology_type homology_gene_stable_id homology_protein_stable_id  \\\n",
       "9044  ortholog_one2one      ENSAMEG00000011292         ENSAMEP00000011868   \n",
       "\n",
       "            homology_species  homology_identity  dn  ds  goc_score  \\\n",
       "9044  ailuropoda_melanoleuca               65.0 NaN NaN      100.0   \n",
       "\n",
       "      wga_coverage  is_high_confidence  homology_id  \n",
       "9044           NaN                 1.0     83864229  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df.gene_stable_id == \"ENSG00000088782\") & (df.homology_gene_stable_id == \"ENSAMEG00000011292\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a301efcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>protein_stable_id</th>\n",
       "      <th>species</th>\n",
       "      <th>identity</th>\n",
       "      <th>homology_type</th>\n",
       "      <th>homology_gene_stable_id</th>\n",
       "      <th>homology_protein_stable_id</th>\n",
       "      <th>homology_species</th>\n",
       "      <th>homology_identity</th>\n",
       "      <th>dn</th>\n",
       "      <th>ds</th>\n",
       "      <th>goc_score</th>\n",
       "      <th>wga_coverage</th>\n",
       "      <th>is_high_confidence</th>\n",
       "      <th>homology_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3977993</th>\n",
       "      <td>ENSG00000163959</td>\n",
       "      <td>ENSP00000296327</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>84.4118</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>ENSAMEG00000003867</td>\n",
       "      <td>ENSAMEP00000004087</td>\n",
       "      <td>ailuropoda_melanoleuca</td>\n",
       "      <td>84.9112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83859110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_stable_id protein_stable_id       species  identity  \\\n",
       "3977993  ENSG00000163959   ENSP00000296327  homo_sapiens   84.4118   \n",
       "\n",
       "            homology_type homology_gene_stable_id homology_protein_stable_id  \\\n",
       "3977993  ortholog_one2one      ENSAMEG00000003867         ENSAMEP00000004087   \n",
       "\n",
       "               homology_species  homology_identity  dn  ds  goc_score  \\\n",
       "3977993  ailuropoda_melanoleuca            84.9112 NaN NaN      100.0   \n",
       "\n",
       "         wga_coverage  is_high_confidence  homology_id  \n",
       "3977993           NaN                 1.0     83859110  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df.gene_stable_id == \"ENSG00000163959\") & (df.homology_gene_stable_id == \"ENSAMEG00000003867\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d7efd91",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (compara)",
   "language": "python",
   "name": "compara"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
